minimal-bootstrap: use recursive FOD to make nix unpack bootstrap sources

emilytrau opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

This is an addendum to work done in #238357. Investigating it seems like the contents of the narball wasn't being unpacked, causing hex0 to not find sources for hex1 and hang. Switching to a recursive FOD seems like a solution?

Related #227914

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

emilytrau committed:
minimal-bootstrap: use recursive FOD to make nix unpack bootstrap sources

amjoseph-nixpkgs commented on pkgs/os-specific/linux/minimal-bootstrap/stage0-posix/bootstrap-sources.nix:
```diff
@@ -85,22 +85,13 @@ rec {
       #
-      #  ${name}.nar.xz
+      #  /nix/store/________________________________-${name}
       #
-      # Please obtain or create this file, give it exactly the name
-      # shown above, and then run the following command:
-      #
-      #   nix-store --add-fixed ${outputHashAlgo} ${name}.nar.xz
-      #
-      # You can create this file from an already-bootstrapped nixpkgs
+      # You can create this path from an already-bootstrapped nixpkgs
       # using the following command:
       #
       #   nix-build '<nixpkgs>' -A make-minimal-bootstrap-sources
       #
-      # Or, if you prefer, you can create this file using only `git`,
-      # `nix`, and `xz`.  For the commands needed in order to do this,
-      # see `make-bootstrap-sources.nix`.
-      #
```
Being able to give the user these instructions is kinda important.  Let me see if I can find a way to get back to that point.

Ericson2314 replied:
We can have the same instructions for doing it manually with the other Nix store add command, no?

amjoseph-nixpkgs replied:
That's what I'm going to try to figure out.

Ericson2314 replied:
I added them bakc.

Ericson2314 committed:
Address my comments

1. Advice for manually assembling the bootstrap sources is restored.

2. Hash is no longer duplicated

3. More informative store path using `builtins.placeholder` is provided.

Ericson2314 commented:
It builds again, but unfortunately I get a segfault building glibc.

zimbatm commented:
I'm not sure we fully understand the implications of this change. It's effectively changing the position of cache.nixos.org from being a cache to being a source for bootstrapping nixpkgs.

The advantage is pretty obvious, but I think there is a downside where bootstrapping becomes harder if cache.nixos.org were to disappear or lost those files.

Ericson2314 commented:
I would say it's probably better to worry about such things when we're ready to switch over to this.

tarballs.nixos.org can be a substituter for fixed output things with a little bit of work, native git hashing support should hopefully make fetching these sources easier, etc.

zimbatm commented:
Isn't `nix-build '<nixpkgs>' -A make-minimal-bootstrap-sources` relying on the previous tarball to build itself?

amjoseph-nixpkgs commented:
> Isn't `nix-build '<nixpkgs>' -A make-minimal-bootstrap-sources` relying on the previous tarball to build itself?

Keep reading....

https://github.com/emilytrau/nixpkgs/blob/4179372632f8f5313ac8684690a0f636295e9146/pkgs/os-specific/linux/minimal-bootstrap/stage0-posix/bootstrap-sources.nix#L95-L97

Of course, it would be nice if we actually *printed* those lines, so the user could copy-paste them into the terminal.
